### PR TITLE
code cleanup

### PR DESCRIPTION
**Pull request by tnawathe21**
## Description
Summary of changes:
- Dealt with all build warnings -- this is important because the actual envoy bazel configuration treats warnings as errors
- Verified that no status/result variables were being reused as mentioned [here](https://github.com/DataDog/envoy-filter-example/pull/35#discussion_r1295966503)
- Changed error codes to be more descriptive
- Used `std::move` alongside `std::string` whenever possible to avoid unnecessary copies
- Removed unneeded protobuf field
- Renamed filter from `sample` to `header_rewrite`
- Verified behavior for edge cases, eg. extra whitespace
- Added checks to verify that no `operation_expression` is being accessed out of bounds
- Made variables `const` whenever possible
- Removed metadata integration test (see #40)

JIRA=[EDGEBE-467](https://datadoghq.atlassian.net/browse/EDGEBE-467)

## Testing
Added to and re-ran unit tests

[EDGEBE-467]: https://datadoghq.atlassian.net/browse/EDGEBE-467?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by tnawathe21** [comment]
Current dependencies on/for this PR:
* main
  * **PR #37** <a href="https://app.graphite.dev/github/pr/DataDog/envoy-filter-example/37" target="_blank">Graphite</a> 
    * **PR #38** <a href="https://app.graphite.dev/github/pr/DataDog/envoy-filter-example/38" target="_blank">Graphite</a>  👈
      * **PR #39** <a href="https://app.graphite.dev/github/pr/DataDog/envoy-filter-example/39" target="_blank">Graphite</a> 
        * **PR #40** <a href="https://app.graphite.dev/github/pr/DataDog/envoy-filter-example/40" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/DataDog/envoy-filter-example/38?utm_source=stack-comment).

File: header-rewrite-filter/header_rewrite_config.cc [hunk 1]
```diff
@@ -30,7 +30,7 @@ class HttpHeaderRewriteFilterConfigFactory : public NamedHttpFilterConfigFactory
     return ProtobufTypes::MessagePtr{new envoy::extensions::filters::http::HeaderRewrite()};
   }
 
-  std::string name() const override { return "sample"; }
+  std::string name() const override { return "envoy.header_rewrite"; }
```
Comment (rob05c) [inline header-rewrite-filter/header_rewrite_config.cc:33]: Nit: would be good to put this in a const in the header

**Review by rob05c — APPROVED** [review]
LGTM, comment is minor

File: header-rewrite-filter/header_rewrite_config.cc (hunk 1 above)
Comment (henrymwang) [reply header-rewrite-filter/header_rewrite_config.cc:33]: I don't have a strong opinion either way, but other extensions just inline it https://github.com/envoyproxy/envoy/blob/main/source/extensions/health_checkers/thrift/config.h#L24

**Review by henrymwang — APPROVED** [review]
